### PR TITLE
optimized Refrences hint paths

### LINKED ISSUE #663: better dll refrence paths
Currently we need to copy reference paths to `Dependencies/`  folder in order to built this mod.
This means that we have duplicated version of each library and sometimes both get loaded by de-cpompilers slowing them down.
To add insult to injury the libraries are then copied into the bin folder creating triple version of each library.

One is good enough. It would be better to instruct .csproj to look into the application installation folder first (using`ProgramFiles` macro) -if that failed then we can look into `Dependencies/`. The copy local should also be set to false. Something like this:
```
    <Reference Include="Assembly-CSharp">
      <HintPath>$(ProgramFiles)\Steam\steamapps\common\Cities_Skylines\Cities_Data\Managed\Assembly-CSharp.dll</HintPath>
      <HintPath>~\Library\Application Support\Steam\SteamApps\common\Cities_Skylines\Cities_Data\Managed\Assembly-CSharp.dll</HintPath>
      <HintPath>..\dependencies\Assembly-CSharp.dll</HintPath>
      <Private>False</Private>
    </Reference>
```
EDIT: this way it also makes the job of first time user developers easier.

### PR DESCRIPTION
**Pull request by kianzarrin**
Fixes #663
I fixed this only for Assembely-csharp. if you agree with my approach I will do this to all CS dlls across all projects.

EDIT: modified the editor config so that csproj would consistently use 2 indent spaces like the rest of the file. It was driving me crazy!

### Test:
- delete the dependencies folder
- Cleane cache : EDIT: clean build - unload and reload project. 
- RESULT: ICities.dll reference is valid proving that the first steam path property is working (other CS dll references are invalid)
- modify first steam path to an invalid path
- RESULT: ICities.dll reference is valid proving that getting steam path from registry was successful.
- modify registry path to something invalid
- RESULT: ICities.dll reference is invalid as expected
- revive dependencies folder
- RESULT: ICities.dll reference is valid proving dependencies is accessable
- revert TLM.csproj
- RESULT: ICities.dll reference is valid and is using C:\Program Files (x86)\Steam\steamapps\common\Cities_Skylines\Cities_Data\Managed\ICities.dll proving this is the prefered path

EDIT: ANYONE USING MacOS HERE? can you please confirm it works on Mac?

Any special setup you have please let me know and I may be able to come up with something.

### CODE REVIEW

**Comment by kianzarrin** [comment]
@kvakvs where Have you installed you Cities skylines? maybe I could come up with a solution something.

I have been searching the internet for a solution to find steam installation path. best I could come up with is $(ProgramFIles) macro. there is also MSBuildProgramFiles[32] but didn't work out for me.

**Comment by aubergine10** [comment]
Is there a way to interact with registry? `HKEY_CURRENT_USER\Software\Valve\Steam`

That will get you something like `C:\Program Files (x86)\Steam\SteamApps`

**Comment by aubergine10** [comment]
https://stackoverflow.com/questions/1905264/how-do-you-read-a-registry-value-using-a-custom-msbuild-task

**Comment by kianzarrin** [comment]
> @aubergine10 Is there a way to interact with registry? HKEY_CURRENT_USER\Software\Valve\Steam
YES! This is it! I have been searching for this for a month! its nice to be part of a team :)

**Comment by aubergine10** [comment]
Docs might need updating too; example: **Environment > Dependencies** section of [/docs/BUILDING_INSTRUCTIONS.md ](https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/blob/master/docs/BUILDING_INSTRUCTIONS.md)

File: TLM/.editorconfig [hunk 1]
```diff
@@ -2,11 +2,9 @@ root=true
```
Comment (aubergine10) [inline TLM/.editorconfig]: Did you intend to include `.editorconfig` in the PR?

File: TLM/.editorconfig [hunk 2]
```diff
@@ -2,11 +2,9 @@ root=true
 
 [*.cs]
 max_line_length=100
-tab_width=8
 insert_final_newline=true
 trim_trailing_whitespace=true
 indent_style=space
-indent_size=4
```
Comment (aubergine10) [inline TLM/.editorconfig]: If this file is intentional, the `indent_size = 4` should be retained unless you add `.cs` to the big wildcard clause around line 99.

File: TLM/TLM/TLM.csproj [hunk 3]
```diff
@@ -78,21 +78,28 @@
     <ErrorReport>prompt</ErrorReport>
     <CodeAnalysisRuleSet>..\TMPE.ruleset</CodeAnalysisRuleSet>
   </PropertyGroup>
+  <PropertyGroup>
+    <SteamPath >~/Library/Application Support/Steam/</SteamPath>
+    <SteamPath Condition="! Exists ('$(SteamPath)')">$(ProgramFiles)\Steam\</SteamPath>
+    <SteamPath Condition="! Exists ('$(SteamPath)')">$(Registry:HKEY_CURRENT_USER\Software\Valve\Steam@SteamPath)</SteamPath>
+    <CSDLLPath>$(SteamPath)\steamapps\common\Cities_Skylines\Cities_Data\Managed</CSDLLPath>
```
Comment (aubergine10) [inline TLM/TLM/TLM.csproj]: Suggest calling this `<ManagedDLLPath>` or something like that (regular modders will instantly know what it means as they are used to working with managed dlls)

File: TLM/TLM/TLM.csproj [hunk 4]
```diff
@@ -555,9 +562,9 @@ xcopy /y "$(TargetDir)0TMPE.Harmony.dll" "%25DEPLOYDIR%25"
 echo THE ASSEMBLY VERSION IS: @(VersionNumber) created at %25time%25
             
 set DEPLOYDIR=
-    </PostBuildEvent>
+</PostBuildEvent>
```
Comment (aubergine10) [inline TLM/TLM/TLM.csproj]: Whitespace issue?

File: TLM/.editorconfig (hunk 1 above)
Comment (kianzarrin) [reply TLM/.editorconfig]: yes. It was driving me crazy!

File: TLM/.editorconfig (hunk 2 above)
Comment (kianzarrin) [reply TLM/.editorconfig]:
```
[*.cs]
max_line_length=100
insert_final_newline=true
trim_trailing_whitespace=true
indent_style=space

[*]
charset=utf-8
end_of_line=crlf
indent_style=space
indent_size=4  <----- LOOK HERE
```
[*] wild card is already taking care of it. Don't worry I tried it and it works :)

File: TLM/TLM/TLM.csproj (hunk 3 above)
Comment (kianzarrin) [reply TLM/TLM/TLM.csproj]: You probably have noticed I am not good at naming.

**Comment by kianzarrin** [comment]
Apart from minor stuff does everybody agree with my ***approach***
should I go ahead and modify all dlls in all .csproj files?

File: TLM/.editorconfig [hunk 5]
```diff
@@ -87,7 +85,8 @@ indent_size=4
 
 [{*.csproj,*.ccproj,*.androidproj}]
 indent_style=space
-indent_size=4
+indent_size=2
```
Comment (kvakvs) [inline TLM/.editorconfig]: Indent size 4
Tab width 8

**Comment by aubergine10** [comment]
> Apart from minor stuff does everybody agree with my _**approach**_
> should I go ahead and modify all dlls in all .csproj files?

I think yes, modify them. Even if it only improves life a little for people building on Windows with fairly normal setup (all steam games in one main `steamapps` folder) it will still be a great improvement as I suspect that will be majority of devs.

For anyone with more unsual setups they can just use the old _documented_ approach of creating a dependencies folder or manually adding reference paths to projects. I'll update the docs to say "try doing a build first and if it fails, you need these extra steps" = pile of crufty config avoided for most people as It Just Works™ for most of them.

**Comment by kianzarrin** [comment]
> For anyone with more unsual setups they can just use the old _documented_ approach of creating a dependencies folder or manually adding reference paths to projects. I'll update the docs to say "try doing a build first and if it fails, you need these extra steps" = pile of crufty config avoided for most people as It Just Works™ for most of them.

good idea. Although it would be more fun to fix it properly. I have some ideas but I haven't get too far because I don't know XML language at all (I learnt some stuff in the past few days for the sake of this review)

**Comment by aubergine10** [comment]
Maybe create a ticket for more advanced stuff then we can see if we can add that in separate PR later?

IMO it's already much better with the cleaning up of hint paths, one of them being the most commonly used location for the managed dlls.

I've got details of how to fix Github Desktop now (Unix-like executables can't deal with Madatory ASLR \*sigh\*) so I'll be doing a fresh clone to desktop just like a new user would. :)

**Comment by aubergine10** [comment]
Is this ready for merging? I'm about to get github back up and running :)

I think there was just the rename of `<CSDLLPath>` to `<ManagedDLLPath>` remaining?

**Comment by kianzarrin** [comment]
> Is this ready for merging? I'm about to get github back up and running :)

Wait a moment I only supported ICities.dll . OK I will apply these changes all across the projects. 

Also I will added special path for @kvakvs . simple hackish work around.

**Comment by kvakvs** [comment]
That is not right to hardcode my paths into Github source.
I can do the old fashioned `Dependencies` directory way.

**Comment by kianzarrin** [comment]
I deleted my dependencies folder. I checked all paths are correct and copy local is false.
Clean build results:
```
========== Build: 9 succeeded, 0 failed, 0 up-to-date, 0 skipped ==========
```

File: TLM/.editorconfig (hunk 5 above)
Comment (aubergine10) [reply TLM/.editorconfig]: tab should be same as indent, both set to `4`?

**Comment by aubergine10** [comment]
Something borked the OptionsFramework dependency in that commit

**Comment by kvakvs** [comment]
Tab width is how real tabs `\t` are displayed when a file with tab is opened. Tabs are always 8, only few codebases assume different width, and devs in those projects are doomed to an eternity of suffering.

**Comment by aubergine10** [comment]
> Tab width is how real tabs `\t` are displayed when a file with tab is opened. Tabs are always 8, only few codebases assume different width, and devs in those projects are doomed to an eternity of suffering.

Updated my review comment

**Comment by kianzarrin** [comment]
finally! appveyer  is happy!

**Comment by krzychu124** [comment]
Yeah, because your `refrences` branch was light years behind `master`😄

**Comment by kianzarrin** [comment]
```
* Or, create a `\TLM\dependencies` symlink to the `...\Managed` folder shown above
```
I wish I had seen this earlier!

**Comment by kianzarrin** [comment]
@aubergine10 I changed tab to 8.
*,csproj indent is 2. The .csproj file is using indent 2 everywhere. Its very difficult to modify it with .editorconfig indent=4

**Comment by kvakvs** [comment]
I am happy with `.csproj` being anything that works for you
Please don't modify indents in `.cs` and `*`

**Comment by kianzarrin** [comment]
I resolved all raised issues.
lets merge this.

**Comment by kianzarrin** [comment]
BTW clean build, unload/reloading projects may be necessary.

**Comment by aubergine10** [comment]
We should provide some info on how to do clean build and uload/load projects.

**Comment by kianzarrin** [comment]
@aubergine10 
> Suggestion: Can we echo the hint path to output?

You can see the hint path by selecting the reference from project window.
Screenshot (469)

> We should provide some info on how to do clean build and uload/load projects.
> Steps to reproduce:
    Delete local repo on disk
    Fresh clone to desktop
    Switched to this branch in git
    Load solution in to VS2019
    Unload and reload all projects
    Clean the build
    Build all

- No need for a clean build if you have a fresh version of repo. Clean build is only to get rid of references that have been copied locally. Particularly now that the copy local is false clean build is even less necessary.
- No need to unload/load projects if you are building. Only if you need the references for auto-complete. Closing and opening the IDE should have the same effect (AFAIK).

I am surprised Appveyer  and I could build the TMPE but you couldn't I will download a fresh version and try it again.

**Comment by kianzarrin** [comment]
@aubergine10 
I don't know what IDE you use but when you switch to my reference branch from master, you need to update sub-modules:
Screenshot (470)

I managed to reproduce your build errors when I did not update sub-modules. After updating sub-modules it works:
Screenshot (471)

Notes:
When we merge this branch into master there is no more need to update submodules for a fresh checkout because you are not changing branch (unless if you change to an old branch).

when members of our dev team pull the new changes they will face the following situation:
- if they fail to update sub-modules nothing happens because the have ../dependencies where their old version of sub-modules is trying to get dlls from
- even if they delete  ../dependencies their old sub-modules will be fine because it has copied the references locally.
- only when they delete ../dependencies AND they perfrom a `clean solution` operation they will get errors which they can resolve by updating sub-modules.
- it probably will not escape their attention to update sub-modules because when you want to commit or when you want to switch to another branch the IDE will complain about out of date sub-modules.

Recommendation: send a memo to update sub-modules but it probably is not a big deal.

**Comment by aubergine10** [comment]
Will test again in a later today - currently got about 6 billion windows and tabs open working on mod compatibility checker

**Comment by krzychu124** [comment]
> I am surprised Appveyer and I could build the TMPE but you couldn't

Answer is quite simple. Appveyor has `dependencies` folder and it recreates whole repository from scratch(including submodules fetch), every time it needs to build a branch

File: TLM/.editorconfig [hunk 6]
```diff
@@ -97,7 +96,7 @@ indent_size=4
 indent_style=space
 indent_size=4
 
-[*.{appxmanifest,asax,ascx,aspx,build,config,cshtml,csproj,dbml,discomap,dtd,htm,html,jsproj,lsproj,master,njsproj,nuspec,proj,props,razor,resw,resx,skin,StyleCop,targets,tasks,vb,vbproj,xaml,xamlx,xml,xoml,xsd}]
+[*.{appxmanifest,asax,ascx,aspx,build,config,cshtml,dbml,discomap,dtd,htm,html,jsproj,lsproj,master,njsproj,nuspec,proj,props,razor,resw,resx,skin,StyleCop,targets,tasks,vb,vbproj,xaml,xamlx,xml,xoml,xsd}]
 indent_style=space
 indent_size=4
-tab_width=4
```
Comment (kianzarrin) [inline TLM/.editorconfig]: This tab with was 4 I changed it to 8. @aubergine10 said that tabwidth = 4 is dangerous.

**Comment by kianzarrin** [comment]
@kvakvs if you pay attention carefully and read the whole .editorconfig you will understand that all tabwidth and indents are as it should be. Can I merge this?

**Comment by kianzarrin** [comment]
sorry aubergine I didn't meant to re-request review from you.

**Comment by kvakvs** [comment]
ok attention's been paid.

**Comment by aubergine10** [comment]
Additional approvals required to merge this - @kvakvs @krzychu124 can you review pls?

**Comment by kvakvs** [comment]
It did not work on my machine due to Steam and Cities installed in non-conventional locations, and i asked not to hardcode my paths into the PR. So cannot say whether it works or not. If it works for others, i'm happy to approve.

**Comment by kianzarrin** [comment]
@kvakvs can you press approve so that I can merge this?

**Comment by kianzarrin** [comment]
@krzychu124 can you approve this so that I can merge please?

**Pull request merged**